**Видео 2**

**Задача 1**

Даны векторы $\vec{a} = \begin{bmatrix}
    1 \\ 2 \\ 3
\end{bmatrix}$,
$\vec{b} = \begin{bmatrix}
    -5 \\ -1 \\ 6
\end{bmatrix}$, $\vec{c} = \begin{bmatrix}
    -1 \\ 3 \\ 9
\end{bmatrix}$.

Найдите результат вычисления: $3\vec{a} + 2\vec{b} - 4\vec{c}$.

In [88]:
import sympy as sp

a = sp.Matrix([[1], [2], [3]])
b = sp.Matrix([[-5], [-1], [6]])
c = sp.Matrix([[-1], [3], [9]])


3*a + 2*b - 4*c

Matrix([
[ -3],
[ -8],
[-15]])

**Задача 2**

Даны точки $\mathbf{A}(3,-2)$ и $\mathbf{B}(1,4)$. Точка $\mathbf{M}$ лежит на прямой $\mathbf{AB}$ так, что
$|\mathbf{AM}| = 3|\mathbf{AB}|$. Найдите координаты точки $ \mathbf{M}$, если точки $ \mathbf{M}$ и $ \mathbf{B}$ лежат по одну сторону от A.

Решение

Найдём вектор $\mathbf{AB}$: 

B - A = (1 - 3), (4 + 2) = (-2, 6)

Предположу, что из условия $|\mathbf{AM}| = 3|\mathbf{AB}|$ можно найти вектор $\mathbf{AM}$ следующим образом:

AM = 3 * AB = 3 * (-2, 6) = (-6, 18)

Далее найдём координаты точки $\mathbf{M}$:

AM = M - A = (x - 3), (y + 2) => x-3 = -6; y + 2 = 18 => x = -3; y = 16

Координаты точки $\mathbf{M}$ равны $\bf(-3, 16)$




In [89]:
#Посчитаем скалярное произведение векторов AB и AM, чтобы убедиться, что точка М лежит по одну сторону с B

b = sp.Matrix([[-2], [6]])
m = sp.Matrix([[-6], [18]])

sp.Matrix.dot(b, m)

#скалярное произведение положительно, значит точки B и M лежат по одну сторону от А

120

**Видео 4**

**Задача 1**

Найдите векторное произведение двух векторов:
    \begin{align*}
        \vec{a} = \begin{bmatrix}-5\\3\\8\end{bmatrix},\ \vec{b} = \begin{bmatrix}2\\5\\1\end{bmatrix} \\
         \end{align*}

In [90]:

a = sp.Matrix([[-5], [3], [8]])
b = sp.Matrix([[2], [5], [1]])

sp.Matrix.cross(a, b)

Matrix([
[-37],
[ 21],
[-31]])

**Задача 2**

Упростите выражение:
    \begin{align*}
        (\vec{a} + \vec{b} ) \times (\vec{a} - \vec{b})
    \end{align*}



In [91]:
from sympy import MatrixSymbol, Matrix

a = MatrixSymbol('a', 3, 1)
b = MatrixSymbol('b', 3, 1)

vl = (a + b)
vp = (a - b)


r1 = sp.Matrix.cross(Matrix(vl), Matrix(vp))
r1


Matrix([
[-(a[1, 0] - b[1, 0])*(a[2, 0] + b[2, 0]) + (a[1, 0] + b[1, 0])*(a[2, 0] - b[2, 0])],
[ (a[0, 0] - b[0, 0])*(a[2, 0] + b[2, 0]) - (a[0, 0] + b[0, 0])*(a[2, 0] - b[2, 0])],
[-(a[0, 0] - b[0, 0])*(a[1, 0] + b[1, 0]) + (a[0, 0] + b[0, 0])*(a[1, 0] - b[1, 0])]])

In [ ]:
r2 = sp.simplify(r1)
r2


Matrix([
[-2*a[1, 0]*b[2, 0] + 2*a[2, 0]*b[1, 0]],
[ 2*a[0, 0]*b[2, 0] - 2*a[2, 0]*b[0, 0]],
[-2*a[0, 0]*b[1, 0] + 2*a[1, 0]*b[0, 0]]])

**Задача 3**

Все три вектора $\vec{a}, \vec{b},\vec{c}$ имеют длину $3$, и $\vec{a} + \vec{b} + \vec{c} = \vec{0}$.

Найдите результат вычисления: $\vec{a} \cdot \vec{b} + \vec{b} \cdot \vec{c} + \vec{c} \cdot \vec{a}$.

*Подсказка: необходимо нарисовать этот случай (треугольник)*.

Т.к. длина 3х векторов одинакова и равна 3, следовательно треугольник равносторонний, а у равностороннего треугольника углы равны 120 градусов.
cos 120 градусов равен -1/2.

Разложим искомую формулу:
$\vec{a} \cdot \vec{b} + \vec{b} \cdot \vec{c} + \vec{c} \cdot \vec{a}$ = |a| * |b| * (-1/2) + |b| * |c| * (-1/2) + |c| * |a| * (-1/2) =
-1/2 * 27 = $\bf-27/2$


**Видео 6**

**Задача 1**

Даны $A=\begin{bmatrix} 2 & -1 & -1\end{bmatrix}$ и $B=\begin{bmatrix} -2 \\ -1 \\ 3 \end{bmatrix}$. \\
    1. Найдите $AB$ и $BA$, если они существуют. \\
    2. Найдите $A^TB$ и $BA^T$, если они существуют.

In [118]:

a = sp.Matrix([2, -1, -1]).T #матрица в sympy представлены в столбик по умолчанию (3х1)
b = sp.Matrix([[-2], [-1], [3]])

a*b #1x3 * 3x1 = 3x3

Matrix([[-6]])

In [119]:
b*a #3x1 * 1x3= 1x1

Matrix([
[-4,  2,  2],
[-2,  1,  1],
[ 6, -3, -3]])

In [120]:
#a.transpose() * b   решения не будет, т.к 3x1 * 3*1 размерности не корректные#
# b * a.transpose() решения не будет по аналогичной причине


**Задача 2**

Дана матрица $A = \begin{bmatrix} 1 & 4 \\ 9 & 1\end{bmatrix}$.

Найдите $l$, если определитель $A - El$ равен $0$, где $E$ — единичная матрица.

$El = \begin{bmatrix} l & 0 \\ 0 & l\end{bmatrix}$.

In [121]:
from sympy import Eq, solve
l = sp.Symbol('l')

#напишем матрицу, определитель которой равен 0
a = sp.Matrix([[1 - l, 9], [4, 1 - l]])
#далее получим уравнение через формулу нахождения определителя
eq = sp.det(a)
#и решив его получим результат
s = solve(eq, l)

s

[-5, 7]

**Видео 7**

**Задача 1**

Вычислите ранг матрицы $\begin{bmatrix}1&2\\2&4\\1&1\end{bmatrix}$ с помощью функции $rank()$ и c помощью $rref()$.

In [122]:
a = sp.Matrix([[1, 2], [2, 4], [1, 1]])

sp.Matrix.rank(a)

2

In [123]:
sp.Matrix.rref(a)[0]

Matrix([
[1, 0],
[0, 1],
[0, 0]])

**Видео 8**

**Задача 1**

Решите уравнения:

1) $\begin{bmatrix}2&5\\1&3\end{bmatrix}X=\begin{bmatrix}2&1\\1&1\end{bmatrix}$; \\
2)  $X\begin{bmatrix}2&5\\1&3\end{bmatrix}=\begin{bmatrix}2&1\\1&1\end{bmatrix}$;

In [124]:
m1 = sp.Matrix([[2, 5], [1, 3]])
rez = sp.Matrix([[2, 1], [1, 1]])



#чтобы найти матрицу X нужно умножить результат уравнения на обратную матрицу от первого множителя(для каждого уравнения важен порядок умножения матриц)
m1I = m1.inv()
#Для уравнения №1:
x = m1I * rez
x


Matrix([
[1, -2],
[0,  1]])

In [125]:
#проверка
m1 * x 

Matrix([
[2, 1],
[1, 1]])

In [126]:
#Для уравнения №2
x = rez*m1I
x

Matrix([
[5, -8],
[2, -3]])

In [127]:
#Проверка
x*m1

Matrix([
[2, 1],
[1, 1]])

**Задача 2**

Найдите значения $AA^+$, $AA^+A$ и $A^+A$, если $A=\begin{bmatrix}3 \\ 7 \\ 2\end{bmatrix}$.

In [128]:
a = sp.Matrix([3,7,2])
aPI = a.pinv()

aPI

Matrix([[3/62, 7/62, 1/31]])

In [129]:
a*aPI

Matrix([
[ 9/62, 21/62, 3/31],
[21/62, 49/62, 7/31],
[ 3/31,  7/31, 2/31]])

In [130]:
a*aPI*a

Matrix([
[3],
[7],
[2]])

In [131]:
aPI*a

Matrix([[1]])

**Видео 9**

**Задача 1**

Решите уравнения:

1) $x^2 + 2x + 4 =0$ \\
2) $\bar{x} = 2 + 2x$

In [132]:
#уравнение №1
x = sp.Symbol('x')

eq = x**2 + 2*x + 4

solve(eq, x)

[-1 - sqrt(3)*I, -1 + sqrt(3)*I]

In [133]:
#уравнение №2
eq = sp.conjugate(x) - 2 - 2*x

solve(eq, x)

[-2]